Add release note and update doc regarding ICU (#14712)
matrix-org/synapse: docs/upgrade.md

    dpkg -i matrix-synapse-py3_1.3.0+stretch1_amd64.deb
    ```

# Upgrading to v1.74.0

## Unicode support in user search

This version introduces optional support for an [improved user search dealing with Unicode characters](https://github.com/matrix-org/synapse/pull/14464).

To do so it uses the ICU library through PyICU.

> If you want to take advantage of this feature you need to install PyICU

Reviewer: note that this is only for people installing from source or PyPI — Docker and Debs not affected.

(can we find out if we include the dependency in docker and debs and say so here?)

Author: Hummm we do but I think it's not working: we install the dev package in the build image, but we don't install the runtime package in the runtime image. I'll fix that in an independent PR.

Author: Actually it does because it's a dependency of one of our dependency. I'll add an explicit install in case of.

Author: For deb `shlibs:Depends` is taking care of that I believe.

Author: I've sneak the Dockerfile change here since it is not a real fix but more a `let's be explicit`.

> To do so it uses the ICU library through PyICU.
> If you want to take advantage of this feature you need to install PyICU
> but also the ICU native dependency and its development headers
> so that PyICU can build since no prebuilt wheels are available.

Reviewer 2: ```suggestion
If you want to take advantage of this feature you need to install PyICU,
the ICU native dependency and its development headers
so that PyICU can build since no prebuilt wheels are available.
```

> but also ICU native dependency and its development headers
> so that PyICU can build since no prebuild wheels is available.
> You can follow [PyICU documentation](https://pypi.org/project/PyICU/) to do so.

Reviewer: ```suggestion
but also the ICU native dependency and its development headers
so that PyICU can build since no prebuilt wheels are available.
You can follow [PyICU documentation](https://pypi.org/project/PyICU/) to do so.
```

> 

Reviewer: mention that you need to specify e.g. `pip install matrix-synapse[icu]` when installing from PyPI

> You can follow [PyICU documentation](https://pypi.org/project/PyICU/) to do so.

Reviewer 2: ```suggestion
You can follow [the PyICU documentation](https://pypi.org/project/PyICU/) to do so.
```

    dpkg -i matrix-synapse-py3_1.3.0+stretch1_amd64.deb
    ```

# Upgrading to v1.74.0

## Unicode support in user search

This version introduces optional support for an [improved user search dealing with Unicode characters](https://github.com/matrix-org/synapse/pull/14464).

If you want to take advantage of this feature you need to install PyICU,
the ICU native dependency and its development headers
so that PyICU can build since no prebuilt wheels are available.

You can follow [the PyICU documentation](https://pypi.org/project/PyICU/) to do so,

> and then do `pip install matrix-synapse[icu]` for a PyPI install.

Reviewer 3: Shouldn't this have been `matrix-synapse[user-search]` instead of `matrix-synapse[icu]`? The latter did not work for me because `user-search` is the name of the extra in the code (see [here](https://github.com/matrix-org/synapse/pull/14464/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R234-R237)) and using that did work for me

Reviewer 4: Yes, good spot. https://github.com/matrix-org/synapse/pull/14797
